File: inst/htmlwidgets/progressBar.js
HTMLWidgets.widget({

  name: 'progressBar',

  type: 'output',

  factory: function(el, width, height) {

    // TODO: define shared variables for this instance

    return {

      renderValue: function(x) {

        if(x.colorTo){
          var colorTo = x.colorTo;
        } else {
          var colorTo = x.color;
        }

        if(x.colorFrom){
          var colorFrom = x.colorFrom;
        } else {
          var colorFrom = x.color;
        }

        if(x.type == 'line'){
          var progress = new ProgressBar.Line('#' + el.id, {
            strokeWidth: x.strokeWidth,
            easing: x.easing,
            duration: x.duration,
            color: x.color,
            trailColor: x.trailColor,
            trailWidth: x.trailWidth,
            svgStyle: {width: x.svgWidth, height: x.svgHeight},
            from: {color: colorFrom},
            to: {color: colorTo},
            step: (state, bar) => {
                bar.path.setAttribute('stroke', state.color);
            }
          });
        } else if(x.type == 'circle'){
          var progress = new ProgressBar.Circle('#' + el.id, {
            strokeWidth: x.strokeWidth,
            easing: x.easing,
            duration: x.duration,
            color: x.color,
            trailColor: x.trailColor,
            trailWidth: x.trailWidth,
            from: {color: colorFrom},
            to: {color: colorTo},
            step: (state, bar) => {
                bar.path.setAttribute('stroke', state.color);
            }
          });
        } else {
          var progress = new ProgressBar.SemiCircle('#' + el.id, {
            strokeWidth: x.strokeWidth,
            easing: x.easing,
            duration: x.duration,
            color: x.color,
            trailColor: x.trailColor,
            trailWidth: x.trailWidth,
            from: {color: colorFrom},
            to: {color: colorTo},
            step: (state, bar) => {
                bar.path.setAttribute('stroke', state.color);
            }
          });
        }

        progress.animate(x.value);

      },

      resize: function(width, height) {

        // TODO: code to re-render the widget with a new size

      }

    };
  }
});

File: R/progressBar.R
#' Set progressbar
#'
#' Create a progressbar
#'
#' @param value percentage: to pass less than \code{1}.
#' @param stroke.width width of stroke.
#' @param easing easing function.
#' @param duration animation duration.
#' @param trail.color color of trail.
#' @param trail.width width of trail.
#' @param color color of bar.
#' @param svg.width,svg.height width and height of svg.
#' @param color.from,color.to starting and ending color (optional) see examples.
#' @param width,height dimensions of container.
#' @param elementId id of container.
#'
#' @import htmlwidgets
#'
#' @export
progressBar <- function(value, stroke.width = 1, easing = "easeInOut", duration = 3500, trail.color = '#eee', trail.width = 1,
                        color = '#ffeb3b', svg.width = "100%", svg.height = "100%", color.from = NULL, color.to = NULL,
                        width = NULL, height = 8, elementId = NULL) {

  # forward options using x
  x = list(
    type = 'line',
    color = color,
    strokeWidth = stroke.width,
    easing = easing,
    duration = duration,
    trailColor = trail.color,
    trailWidth = trail.width,
    svgHeight = svg.height,
    svgWidth = svg.width,
    colorFrom = color.from,
    colorTo = color.to,
    value = value
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'progressBar',
    x,
    width = width,
    height = height,
    package = 'progressBaR',
    elementId = elementId
  )
}

#' Shiny bindings for progressBar
#'
#' Output and render functions for using progressBar within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a progressBar
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name progressBar-shiny
#'
#' @export
progressBarOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'progressBar', width, height, package = 'progressBaR')
}

#' @rdname progressBar-shiny
#' @export
renderProgressBar <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, progressBarOutput, env, quoted = TRUE)
}
